homeassistant/components/recorder/purge.py

```diff
@@ -104,6 +108,22 @@ def _purge_state_ids(session: Session, state_ids: list[int]) -> None:
     _LOGGER.debug("Deleted %s states", deleted_rows)
 
 
+def _evict_purged_state_from_old_states_cache(
+    instance: Recorder, purged_state_ids: list[int]
+) -> None:
+    """Evict purged states from the old states cache."""
+    # Make a map from old_state_id to entity_id
+    old_state_reversed = {}
+    for entity_id, old_state in instance._old_states.items():
+        if old_state.state_id:
+            old_state_reversed[old_state.state_id] = entity_id
+
+    # Evict any purged state from the old states cache
+    old_state_ids = set(old_state_reversed.keys())
+    for purged_state_id in old_state_ids.intersection(purged_state_ids):
```

Comment:
Should we make `_select_state_ids_to_purge` return a set instead of converting it to a set here?

Reply:
Will check if that's safe when I'm not mobile

Reply:
Looks like that should be safe to do.